src/Formik.tsx

```diff
@@ -1,312 +1,232 @@
-    this.fields = {};
-    this.initialValues = props.initialValues || ({} as any);
-    warning(
-      !(props.component && props.render),
-      'You should not use <Formik component> and <Formik render> in the same <Formik> component; <Formik render> will be ignored'
-    );
+  }>({});
+  const [state, dispatch] = React.useReducer<
+    FormikState<Values>,
+    FormikMessage<Values>
+  >(formikReducer, {
+    values: props.initialValues,
+    errors: {},
+    touched: {},
+    isSubmitting: false,
+    isValidating: false,
+    submitCount: 0,
+  });
 
-    warning(
-      !(props.component && props.children && !isEmptyChildren(props.children)),
-      'You should not use <Formik component> and <Formik children> in the same <Formik> component; <Formik children> will be ignored'
-    );
+  const runValidationAsEffect = React.useCallback(
+    () => {
+      const [validate, cancel] = makeCancelable(validateForm(state.values));
+      validate.then(x => x).catch(x => x); // catch the rejection silently
+      return cancel;
+    },
+    [state.values]
+  );
 
-    warning(
-      !(props.render && props.children && !isEmptyChildren(props.children)),
-      'You should not use <Formik render> and <Formik children> in the same <Formik> component; <Formik children> will be ignored'
-    );
-  }
+  React.useEffect(
+    () => {
+      if (!!validateOnChange && !state.isSubmitting) {
+        return runValidationAsEffect();
+      }
+      return;
+    },
+    [state.values, state.isSubmitting, validateOnChange]
+  );
 
-  registerField = (name: string, Comp: React.Component<any>) => {
-    this.fields[name] = Comp;
-  };
+  React.useEffect(
+    () => {
+      if (!!validateOnBlur && !state.isSubmitting) {
+        return runValidationAsEffect();
+      }
+      return;
+    },
+    [state.touched, state.isSubmitting, validateOnBlur]
+  );
 
-  unregisterField = (name: string) => {
-    delete this.fields[name];
+  const imperativeMethods = {
+    resetForm,
+    submitForm,
+    validateForm,
+    validateField,
+    setErrors,
+    setFieldError,
+    setFieldTouched,
+    setFieldValue,
+    setStatus,
+    setSubmitting,
+    setTouched,
+    setValues,
+    setFormikState,
   };
 
-  componentDidMount() {
-    this.didMount = true;
-  }
-
-  componentWillUnmount() {
-    // This allows us to prevent setting state on an
-    // unmounted component. This can occur if Formik is in a modal, and submission
-    // toggles show/hide, and validation of a blur field takes longer than validation
-    // before a submit.
-    // @see https://github.com/jaredpalmer/formik/issues/597
-    // @see https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html
-    this.didMount = false;
-
-    // Cancel validation on unmount.
-    if (this.validator) {
-      this.validator();
+  function registerField(name: string, { validate }: any) {
+    if (fields.current !== null) {
+      fields.current[name] = {
+        validate,
+      };
     }
   }
```

Comment:
I think we might want to wrap all these functions defined in `useFormik` with `useCallback` so that we're not creating new functions on every render.

Reply:
But also with ~26 functions that could be a real pain to manage.

I wonder if it's possible to separate the internals of `useFormik` into logical parts like validation and values.  I might experiment with that and report back.

Reply:
Let’s start with useCallback. Let’s discuss how to slice it up. I agree it’s a lot.

Reply:
Will cut a new release in the meantime